### ISSUE TITLE
Severe bug in ``imshow`` when plotting images with small values

### ISSUE THREAD

**Issue by astrofrog** [matplotlib/matplotlib#6394]
The following minimal example:

```
import matplotlib
matplotlib.use('Agg')
from astropy.wcs import WCS
from wcsaxes import datasets
import matplotlib.pyplot as plt
hdu = datasets.fetch_msx_hdu()
plt.imshow(hdu.data, vmin=-2e-4, vmax=2e-4)
plt.savefig('image.png')
```

produces the following output:

![image](https://cloud.githubusercontent.com/assets/314716/15148599/dedc6636-16bd-11e6-83f8-a06d4c5f5757.png)

This appears to be related to the fact the values in the image data are < 1. If I multiply all the values by 1e6, I get something sensible:

```
plt.imshow(hdu.data * 1e6, vmin=-2e-4, vmax=2e-4)
plt.savefig('image.png')
```

![image](https://cloud.githubusercontent.com/assets/314716/15148648/26fd7162-16be-11e6-84b9-347e388d99f2.png)

**Comment by tacaswell** [comment]
attn @mdboom 

It looks like this is master or 2.x?

**Comment by astrofrog** [comment]
This is on master

**Comment by WeatherGod** [comment]
What is the dtype of the input data? What is its range? I wonder if
something degenerate is happening during normalization.

Also, which version of matplotlib are you using? The master branch has a
completely overhauled image architecture that may or may not suffer from
this.

On Tue, May 10, 2016 at 9:48 AM, Thomas Robitaille <notifications@github.com

**Comment by astrofrog** [comment]
The data is ``dtype('>f8')``, so could be a big-endian issue?

**Comment by jenshnielsen** [comment]
I can reproduce it with 2.x and master but not on 1.5.1 so it's likely to be a result of the image refactoring

The following warning is also raised on 2.x/master

```python
/Users/jhn/Envs/mpl2x/lib/python3.5/site-packages/matplotlib/colors.py:435: RuntimeWarning: overflow encountered in multiply
  xa *= self.N
```

**Comment by tacaswell** [comment]
It is something funny with the `'>f8'` type.  If you do anything to change the data type of the input (multiply by `1`, `hdu.data.astype('float')`, `hdu.data.byteswap().newbyteorder()`) fixes the plot.

@mdboom I think the problem is in here https://github.com/matplotlib/matplotlib/blob/master/src/_image_wrapper.cpp#L280 which does not do anything special about non-native byte order.  I am not sure of the right place to fix this (c++ layer or where we intake the data).

**Comment by efiring** [comment]
This should be fixed by #6422.  Please reopen if you find this is not the case.
